### PR TITLE
Uniting Finnegan's Wake demo logic with current master.

### PR DESCRIPTION
**Pull request by jMyles**
* Updates all of the demo machinery (non-mining node script, middleware, demo script) with new features and semantics
* Cleans up signing logic for node verification
* Allows TreasureMaps to *either* be passed in *or* looked up by ID, instead of needing two steps
* Cleans up some tests in light of #356

### CODE REVIEW

File: examples/run_ursula_with_rest_and_dht_but_no_mining.py [hunk 1]
```diff
@@ -19,14 +19,19 @@
...
-_URSULA = Ursula(dht_port=3501, rest_port=3601, ip_address="localhost", db_name=DB_NAME, federated_only=True)
+_URSULA = Ursula(dht_port=3501,
+                 rest_port=3601,
+                 rest_host="localhost",
+                 dht_host="localhost",
+                 db_name=DB_NAME,
+                 federated_only=True)
 _URSULA.dht_listen()
 
 CURVE = ec.SECP256R1
```
Comment (tuxxy) [inline examples/run_ursula_with_rest_and_dht_but_no_mining.py]: Nice try lizard people.

File: nucypher/characters.py [hunk 2]
```diff
@@ -852,30 +871,27 @@ def retrieve(self, message_kit, data_source, alice_verifying_key):
...
-            raise RuntimeError("Not enough known nodes.  Try following the TreasureMap again.")
+        cleartexts = []
 
-        work_orders = self.generate_work_orders(hrac, message_kit.capsule)
-        for node_id in self.treasure_maps[hrac]:
-            node = self._known_nodes[UmbralPublicKey.from_bytes(node_id)]
-            cfrags = self.get_reencrypted_c_frags(work_orders[bytes(node.stamp)])
+        for work_order in work_orders.values():
+            cfrags = self.get_reencrypted_cfrags(work_order)
             message_kit.capsule.attach_cfrag(cfrags[0])
```
Comment (cygnusv) [inline nucypher/characters.py]: Why just the first CFrag?

File: examples/run_ursula_with_rest_and_dht_but_no_mining.py (hunk 1 above)
Comment (tuxxy) [reply examples/run_ursula_with_rest_and_dht_but_no_mining.py]: The lizard people are innocent. This should be renamed to `TLS_CURVE`, though.

Perhaps an upgrade to `SECP384R1` is possible? If not, leave it with P-256 for TLS compatibility.

File: nucypher/characters.py [hunk 3]
```diff
@@ -684,7 +684,14 @@ def peek_at_treasure_map(self, map_id):
 
         Return two sets: nodes that are unknown to us, nodes that are known to us.
         """
-        treasure_map = self.treasure_maps[map_id]
+        if not treasure_map:
```
Comment (tuxxy) [inline nucypher/characters.py]: I figured out a neat little xor trick for this pattern:

```
if not bool(treasure_map) ^ bool(map_id):
    raise ValueError("You can only pass either a treasure_map or a map_id, pick one.")
```

File: nucypher/characters.py [hunk 4]
```diff
@@ -760,10 +780,10 @@ def make_compass_for_alice(self, alice):
     def construct_policy_hrac(self, verifying_key, label):
         return keccak_digest(bytes(verifying_key) + self.stamp + label)
 
-    def construct_map_id(self, verifying_key, label):
+    def construct_hrac_and_map_id(self, verifying_key, label):
```
Comment (tuxxy) [inline nucypher/characters.py]: :+1:

File: nucypher/characters.py [hunk 5]
```diff
@@ -823,7 +842,7 @@ def generate_work_orders(self, hrac, *capsules, num_ursulas=None):
 
         return generated_work_orders
 
-    def get_reencrypted_c_frags(self, work_order):
+    def get_reencrypted_cfrags(self, work_order):
```
Comment (tuxxy) [inline nucypher/characters.py]: :+1: Good naming catch

File: tests/conftest.py [hunk 6]
```diff
@@ -1,7 +1,8 @@
 from cryptography.hazmat.primitives.asymmetric import ec
 from umbral.config import set_default_curve
+from umbral.curve import SECP256K1
 
-set_default_curve(ec.SECP256K1())
+set_default_curve(SECP256K1)
```
Comment (tuxxy) [inline tests/conftest.py]: :+1:

**Pull request merged**